thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -3737,119 +3737,337 @@ class HashJoinSlaveActivity : public CSlaveActivity, implements IStopInput
+     *  2) Compacts HT row pointers to front of HT memory space (re-purposed as a row array)
+     *  3) Sorts row array by aggregate key field(s)
+     *  4) Clears CAggregateHT's hash table members and passes ownership of row array memory
+     *     to a IRowStream implementation.
+     */
+    IRowStream *getRowStream(bool sorted) override
+    {
+        if (sorted)
         {
-            Linked<RowAggregator> localAggregated;
-        public:
-            IMPLEMENT_IINTERFACE;
-            CRowAggregatedStream(RowAggregator *_localAggregated) : localAggregated(_localAggregated)
+            unsigned iPos = 0;
+            void **rowTable = (void **)table;
+            // re-purposing table memory as row ptr table
+            // 1st compact rows in table to the front.
+            while (iPos != htn)
             {
+                HTEntry &ht = table[iPos];
+                if (ht.row)
+                {
+                    *rowTable = rowAllocator->finalizeRow(ht.size, ht.row, ht.size);
+                    ++rowTable;
+                }
+                ++iPos;
             }
-            // IRowStream impl.
-            virtual const void *nextRow()
+            size32_t remainingSz = ((byte *)&table[htn]) - ((byte *)rowTable);
+            memset(rowTable, 0, remainingSz);
+
+            // sort elements
+            unsigned elems = ((byte *)rowTable-(byte *)table)/sizeof(void *);
+            rowTable = (void **)table;
+            parqsortvec(rowTable, elems, *elementComparer, 0);
+
+            // Implements a IRowStream to walk sorted rows that were previously in HT.
+            class CStream : public CSimpleInterfaceOf<IRowStream>
             {
-                Owned<AggregateRowBuilder> next = localAggregated->nextResult();
-                if (!next) return NULL;
-                return next->finalizeRowClear();
-            }
-            virtual void stop() { }
-        };
-        Owned<IRowStream> localAggregatedStream = new CRowAggregatedStream(localAggTable);
-        if (!distributor)
-            distributor.setown(createHashDistributor(&activity, activity.queryContainer().queryJobChannel().queryJobComm(), mptag, false, NULL, "MERGEAGGS"));
-        Owned<IThorRowInterfaces> rowIf = activity.getRowInterfaces(); // create new rowIF / avoid using activities IRowInterface, otherwise suffer from circular link
-        strm.setown(distributor->connect(rowIf, localAggregatedStream, helperExtra.queryHashElement(), NULL));
-        for (;;)
+                void **rows;
+                const void **rowsPtr;
+                const void **endRows;
+            public:
+                CStream(void **_rows, unsigned numRows) : rows(_rows)
+                {
+                    rowsPtr = (const void **)rows;
+                    endRows = rowsPtr+numRows;
+                }
+                ~CStream()
+                {
+                    ::ReleaseRoxieRow(rows);
```

Comment:
Glad you spotted, if not, another case where HPCC-17289 would have been helpful.